servant-auth-server/src/Servant/Auth/Server/Internal/Class.hs

```diff
@@ -1,49 +1,99 @@
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilyDependencies #-}
 {-# LANGUAGE UndecidableInstances #-}
 module Servant.Auth.Server.Internal.Class where
 
-import Servant.Auth
-import Data.Monoid
+import Data.Proxy
+import GHC.Generics (Generic)
+import Network.Wai (Request)
 import Servant hiding (BasicAuth)
 
+import Servant.Auth
 import Servant.Auth.Server.Internal.Types
 import Servant.Auth.Server.Internal.ConfigTypes
 import Servant.Auth.Server.Internal.BasicAuth
 import Servant.Auth.Server.Internal.Cookie
 import Servant.Auth.Server.Internal.JWT
 
+-- | The result of running the auths 'as'
+data AuthResult auths v
+  = Authenticated v
+  | AuthFailed (ErrorList (AuthsToErrors auths))
+  deriving (Generic, Functor, Traversable, Foldable)
+
+deriving instance (Eq v, Eq (ErrorList (AuthsToErrors auths)))
+  => Eq (AuthResult auths v)
+deriving instance (Show v, Show (ErrorList (AuthsToErrors auths)))
+  => Show (AuthResult auths v)
+
+
+-- | Open type family
+type family AuthsToErrors (auths :: [*]) = (errs :: [*]) | errs -> auths where
+  AuthsToErrors '[] = '[]
+  AuthsToErrors (a ': auths) = AuthError a ': AuthsToErrors auths
+
+-- | HList as a GADT. Named differently to prevent clash with servant's HList
+data ErrorList (as :: [*]) where
+  ENil :: ErrorList '[]
+  ECons :: a -> ErrorList as -> ErrorList (a ': as)
```

Comment:
Would it be simpler to have:

```
ECons :: AuthError a -> ErrorList as -> ErrorList (a ': as)
```

and then not have `AuthsToErrors`? (Assuming, I think, that we move to `AuthError` being an associated data family rather than type family).

Reply:
Yes absolutely. I think I was using the `AuthsToErrors` family just so the HList was given the error types. But since we have injectivity of `AuthError` it's not necessary and this makes far more sense.